core/src/block_sync.rs

```diff
@@ -181,11 +184,34 @@ pub mod message {
                         return;
                     }
 
-                    let mut blocks = block_sync.sumeragi.blocks_after_hash(*hash);
-                    blocks.truncate(block_sync.block_batch_size as usize);
+                    let mut h = 1;
+                    loop {
```

Comment:
Consider separating it out. This looks either as a `while let` loop, but I'm not sure (and that's the problem), or a `for` loop over `h`, with a few premature breaks. 

In either case, splitting this function into smaller chunks, with a lower level of nesting would help.